File: benchmarks/pandas/bench_timedelta_array.py
"""
Benchmark: pd.arrays.TimedeltaArray — create and operate on nullable timedelta arrays.
Outputs JSON: {"function": "timedelta_array", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 50

# Build values: ~10% NaT, small durations (i seconds) to avoid overflow
values = np.array(
    [None if i % 10 == 0 else i * 1_000_000_000 for i in range(SIZE)],  # nanoseconds (1 ns/unit)
    dtype=object,
)
td_values = pd.to_timedelta(values, unit="ns")
fill_value = pd.Timedelta(0)


def run():
    arr = pd.array(td_values, dtype="timedelta64[ns]")

    # Component access
    _ = arr.days
    _ = arr.seconds
    _ = arr.total_seconds()

    # Null checks
    _ = arr.isna()
    _ = ~arr.isna()

    # Aggregation (via numpy)
    valid = td_values[~pd.isna(td_values)]
    _ = valid.sum()
    _ = valid.min()
    _ = valid.max()

    # Fill
    _ = arr.fillna(fill_value)


for _ in range(WARMUP):
    run()

start = time.perf_counter()
for _ in range(ITERATIONS):
    run()
total_ms = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "timedelta_array",
    "mean_ms": total_ms / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total_ms,
}))

File: benchmarks/tsb/bench_timedelta_array.ts
/**
 * Benchmark: TimedeltaArray — create and operate on nullable timedelta arrays.
 * Outputs JSON: {"function": "timedelta_array", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { TimedeltaArray, Timedelta } from "../../src/index.js";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 50;

// Build raw values: ~10% null
const values: (number | null)[] = Array.from({ length: SIZE }, (_, i) =>
  i % 10 === 0 ? null : i * 60_000,
);

const fillValue = Timedelta.fromMilliseconds(0);

function run(): void {
  const arr = TimedeltaArray.from(values);

  // Component access
  void arr.days;
  void arr.hours;
  void arr.totalSeconds;

  // Null checks
  void arr.isna();
  void arr.notna();

  // Aggregation
  void arr.sum();
  void arr.min();
  void arr.max();

  // Fill
  void arr.fillna(fillValue);
}

for (let i = 0; i < WARMUP; i++) run();

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) run();
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "timedelta_array",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
